tests/functional_tests/transfer.py

```diff
@@ -1081,5 +1088,236 @@ def diff_transfers(actual_transfers, expected_transfers):
+        assert receiver_wallet.get_balance().balance == expected_receiver_balance
+
+        # Check a fresh incoming wallet with wallet files saved to disk and encrypted with password
+        receiver_wallet.close_wallet()
+        utils.remove_wallet_files('test2')
+        receiver_wallet.restore_deterministic_wallet(seed = seeds[1], filename = 'test2', password = 'test_password')
+
+        receiver_wallet.set_background_sync(enable = True)
+        receiver_wallet.rescan_blockchain(True)
+        diff_transfers(receiver_wallet.get_transfers(), background_transfers)
+        diff_incoming_transfers(receiver_wallet.incoming_transfers(transfer_type = 'all'), background_incoming_transfers)
+        assert receiver_wallet.get_balance().balance == background_bal
+
+        # Try to disable background sync with the wrong password
+        invalid_password = False
+        try: receiver_wallet.set_background_sync(enable = False, password = 'wrong_password')
+        except: invalid_password = True
+        assert invalid_password
+
+        receiver_wallet.set_background_sync(enable = False, password = 'test_password')
+        diff_transfers(receiver_wallet.get_transfers(), transfers)
+        diff_incoming_transfers(receiver_wallet.incoming_transfers(transfer_type = 'all'), incoming_transfers)
+        assert receiver_wallet.get_balance().balance == expected_receiver_balance
+
+        # Close receiver's wallet while background sync is enabled then reopen
+        receiver_wallet.set_background_sync(enable = True)
+        receiver_wallet.rescan_blockchain(True)
+        diff_transfers(receiver_wallet.get_transfers(), background_transfers)
+        diff_incoming_transfers(receiver_wallet.incoming_transfers(transfer_type = 'all'), background_incoming_transfers)
+        assert receiver_wallet.get_balance().balance == background_bal
+        receiver_wallet.close_wallet()
+        receiver_wallet.open_wallet('test2', password = 'test_password')
+        # It should reopen with spend key loaded and correctly scan all transfers
+        diff_transfers(receiver_wallet.get_transfers(), transfers)
+        diff_incoming_transfers(receiver_wallet.incoming_transfers(transfer_type = 'all'), incoming_transfers)
+        assert receiver_wallet.get_balance().balance == expected_receiver_balance
+        utils.remove_wallet_files('test2')
+
+        # Sanity check against incoming wallet restored at height 0
+        receiver_wallet.close_wallet()
+        receiver_wallet.restore_deterministic_wallet(seed = seeds[1], restore_height = 0)
+        receiver_wallet.refresh()
+        diff_transfers(receiver_wallet.get_transfers(), transfers)
+        diff_incoming_transfers(receiver_wallet.incoming_transfers(transfer_type = 'all'), incoming_transfers)
+        assert receiver_wallet.get_balance().balance == expected_receiver_balance
+
+        # cleanup
+        self.create()
+
```

Comment:
Very comprehensive tests, gives a good confidence everything works.

There is one case that does not get tested, but one that must be almost impossible to unit test because of its inherently random component: Check what happens when a transaction of a third wallet (not sender, not receiver) contains one of the involved enotes as a ring member in a transaction.

Such a transaction will get added as a candidate first, but later ignored as a transaction not belonging to the wallet after all.

Fortunately this case is almost trivial to handle for the wallet. All that is needed is that `wallet2::process_new_transaction` as called in `wallet2::process_background_sync_cache` does not freak out and stops the whole list processing altogether when it sees such a candidate that does not play out, but just ignores it gracefully and returns.

I confirmed that it does by testing manually on testnet. On testnet there are so few transactions happening that if you continue to transact with a third wallet quite soon one of the enotes owned by the wallet in background sync will get used as a decoy.